### DISCUSSION TITLE
How to store state in App?

### DISCUSSION THREAD

**Discussion by chbndrhnns** [Textualize/textual#4107] [Q&A] [answered]
Is there a good place to store context or state within a textual app?  
I’d like to keep some business logic separate from the handler and action methods but want to access it from within these methods.

**Comment by TomJGooding** [comment]
This question is so broad it is difficult to offer any meaningful or practical advice. Could you try narrowing down your question, perhaps with a simple example?

**Reply by chbndrhnns** [reply]
I'd like to inject custom state when creating the app instance. 

Here is a simple example where I am just setting an arbitrary attribute on the app after creating it. 
This breaks typing information as no part of the application is aware of that additional attribute.

Is there a better way to do it? 

```py
import pytest
from textual.app import App, ComposeResult
from textual.widgets import Button


class State:
    def __init__(self):
        self.count = 0


class MyUi(App):
    def compose(self) -> ComposeResult:
        yield Button("Default", id="default")

    def on_button_pressed(self, event: Button.Pressed) -> None:
        self.my_state.count += 1


@pytest.fixture(autouse=True)
def anyio_backend():
    return "asyncio"


async def test_():
    state = State()
    app = MyUi()
    app.my_state = state

    async with app.run_test() as pilot:
        await pilot.click("#default")

    assert state.count == 1
```

**Reply by davep** [reply, marked as answer]
> This breaks typing information as no part of the application is aware of that additional attribute.

Your app is simply a Python class, so you can "declare" that attribute as you generally would with any other Python class: create/assign/type the attribute in `__init__` (just like you did in your `State` class).

**Reply by daddycocoaman** [reply]
@davep Hi, I see this was marked as answered but I don't think it solves the intent here. You can still create/assign/type stuff to the app, but when you call something like `self.app` from the app, it returns an `App[object]` type, and not the actual app (in this case `MyUi` which has an additional attribute). So any use of `self.app` won't see that additional attribute, unless you explicitly type that as `self.app: MyUi`. But this breaks of convenience of trying to access `self.app.my_state` for typing.

Would it be possible to use something like `AppType = NewType(<actual App class name>, App)` when the app contextvar gets set and return AppType instead of `App[object]`?

**Reply by davep** [reply]
The person asking the question marked it as answered, not me. I would assume they were confident that their question was satisfied by the answer(s). 

Feel free to ask your unsatisfied question as a new discussion.

**Reply by daddycocoaman** [reply]
Alternatively, and probably way easier, a `context` dict subclass would also work with a custom `get(name, type, default)` function that also accepts a type, and returns that type. So in this case, something like `self.app.context.get("my_state", int, None)` would be useful.
